AI assisted documentation refactoring (#73)
LLNL/radiuss-shared-ci: docs/sphinx/user_guide/concepts.rst

.. ##
.. ## Copyright (c) 2026, Lawrence Livermore National Security, LLC and
.. ## other RADIUSS Project Developers. See the top-level COPYRIGHT file for
.. ## details.
.. ##
.. ## SPDX-License-Identifier: (MIT)
.. ##

.. _concepts-label:

*************
Core Concepts
*************

This section explains the fundamental concepts behind RADIUSS Shared CI,
helping you understand how the system works before diving into setup and
configuration.

=============================
What is RADIUSS Shared CI?
=============================

RADIUSS Shared CI is a **templated CI infrastructure for GitLab** designed
specifically for LLNL open-source projects hosted on GitHub that need to run
tests on Livermore Computing (LC) systems.

> Rather than each project maintaining its own complex CI configuration, RADIUSS
> Shared CI provides a **shared, reusable framework** that projects can consume
> and customize. This approach:
> 
> - **Reduces maintenance burden** - Bug fixes and improvements benefit all users
> - **Ensures consistency** - Common patterns across RADIUSS projects
> - **Simplifies onboarding** - New projects can adopt proven CI practices quickly
> - **Enables specialization** - Projects focus on their build/test scripts, not CI internals

Reviewer: Here I also think it's a bit too much fluff, I would probably try to condense into something more like 
```suggestion
Rather than each project maintaining its own complex CI configuration, RADIUSS
Shared CI provides a **shared, reusable framework** that projects can consume
and customize. The benefits include:

- Bug fixes and improvements automatically benefit all users
- Common patterns across RADIUSS projects
- New projects start with proven CI practices
- Projects focus on build/test scripts, not CI internals
```

Author: I can't say which one is best: I like that with the initial version I can just read the keywords to get a sense of the claim. "consistency" reads faster and clearer than "Common patterns across RADIUSS projects", and I could always read the sentence if I’m not clear on what's beihnd it.

I think it really depends on the way each one reads this. Admittedly, I'm not in a good position to judge since I know the project too much.

Would you say your proposal is more reader-friendly ?

Reviewer: Hmm I think this one is actually good as is upon re-reading. Especially at the beginning of the page it is nice to highlight some benefits!

  Uses ``flux alloc`` for shared allocations and ``flux run`` for jobs.

Each machine component handles:

- Scheduler-specific allocation commands
- Job execution within allocations
- Resource management
- Reproducer generation for local testing

As a user, you primarily interact through **allocation parameters** (e.g.,
``shared_alloc``, ``job_alloc``) which may still be scheduler specific.
The component handles scheduler details, e.g. how to share allocations.

Template Extension Pattern
===========================

RADIUSS Shared CI components export **job templates** that you extend to
define your specific jobs:

.. code-block:: yaml

   # Component exports .job_on_dane template
   # You extend it with your custom variables

   gcc-build:
     extends: .job_on_dane
     variables:
       SPEC: "gcc@11.0.0"

   clang-build:
     extends: .job_on_dane
     variables:
       SPEC: "clang@14.0.0"

Each template includes:

- **Scheduler commands**: Allocation and job execution
- **Reproducer logic**: Generates commands for local testing
- **Machine tags**: Ensures jobs run on correct runners
- **Customization hooks**: ``.custom_job`` for project-specific setup

> This pattern enables:
> 
> - **DRY principle**: Define once, reuse many times
> - **Consistency**: All jobs follow same structure
> - **Flexibility**: Override only what you need
> 

Reviewer: For me at least, this bit doesn't add much extra info. 

```suggestion
```

      ``matrix-pipeline``, ``dane-pipeline``).

   Base Pipeline
      The ``base-pipeline`` component that provides orchestration templates and
      utilities used by all pipelines.

   Job Template
      A YAML template (e.g., ``.job_on_dane``) that jobs extend to inherit
      common configuration.

   Allocation
      A reservation of compute resources on an HPC machine, managed by a
      scheduler (LSF, SLURM, or Flux).

   Shared Allocation
      A single allocation within which multiple jobs run sequentially or in parallel,
      avoiding repeated allocation wait times.

   RSCI_ Variables
      Internal environment variables prefixed with ``RSCI_`` set by components
      to avoid collisions with GitLab globals.

   Service User
      An LLNL service account used to run CI, providing consistent permissions
      and avoiding individual user quota issues.

   Reproducer
      Shell commands printed in job logs that allow you to reproduce the exact
      CI environment and command locally for debugging.

   Machine Check
      A preliminary job that queries machine availability before triggering
      expensive child pipelines.

   GitHub Token

>       A personal access token with adequate premissions, e.g. ``repo:status``

Reviewer: ```suggestion
      A personal access token with adequate permissions, e.g. ``repo:status``
```

   │                                       │ │
   │  ┌─────────────────────────────────┐  │ │
   │  │ dane-up-check ──→ Available?    │──┼─┤
   │  └─────────────────────────────────┘  │ │
   │           ↓ (if available)            │ │
   │  ┌─────────────────────────────────┐  │ │
   │  │ Child: dane-build-and-test      │  │ │
   │  │  ├─ Component: dane-pipeline    │  │ │
   │  │  ├─ Local: custom-jobs.yml      │  │ │
   │  │  └─ Local: jobs/dane.yml        │  │ │
   │  │                                 │  │ │
   │  │     Jobs run on Dane:           │  │ │
   │  │     ├─ gcc-build ──────────────────┼─┤
   │  │     ├─ clang-build ────────────────┼─┤
   │  │     └─ intel-build ────────────────┼─┤
   │  └─────────────────────────────────┘  │ │
   │                                       │ │
   │  (Similar for lassen, corona, etc.)   │ │
   └───────────────────────────────────────┘ │
                                             │
   ┌─────────────────────────────────────────┘
   │
   │  Status Update: "dane-build-and-test: ✓ Success"
   └──→ GitHub PR shows check results

===================
Next Steps
===================

Now that you understand the core concepts, you can:

> - **New users**: Proceed to :doc:`../getting-started/five-minute-setup` (once created)

Reviewer: Doesn't yet exist?
```suggestion
```

Author: It's coming.
